### PR TITLE
Initial work towards checking const eval rules in miri

### PR DESCRIPTION
**Pull request by oli-obk**
cc #217 

These checks are hard to test, because rustc will do them for us before miri ever runs.

### CODE REVIEW

**Comment by RalfJung** [comment]
I don't entirely understand that `const_env` function. What does that have to do with constant evaluation?

I somewhat expected a flag, which would also make it possible to manually test const_env mode on a given rs file.

**Comment by oli-obk** [comment]
the function returns true if we are in a frame which computes a constant. It also needs a check for checking if any function further up the stack does this.

The idea is that we first start checking miri's evaluation of constants within a miri evaluation of a normal program. Testing ctfe would require a new binary that doesn't process main, as main doesn't fulfill the ctfe requirement of being a const fn.

**Comment by RalfJung** [comment]
> The idea is that we first start checking miri's evaluation of constants within a miri evaluation of a normal program.

Oh, I see.  I didn't catch this.

Btw, I assume this also rules out pointer arithmetic? (Like the stuff I did to make `HashMap` work.)

File: src/operator.rs [hunk 1]
```diff
@@ -151,6 +151,13 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
         let usize = PrimValKind::from_uint_size(self.memory.pointer_size());
         let isize = PrimValKind::from_int_size(self.memory.pointer_size());
         if !left_kind.is_float() && !right_kind.is_float() {
+            if (!left.is_bytes() && !right.is_bytes()) && self.frame().const_env() {
+                if left.is_ptr() && right.is_ptr() {
+                    return Err(EvalError::NotConst("Comparing pointers".to_string()));
```
Comment (RalfJung) [inline src/operator.rs]: "Comparing" sounds wrong here... this could also be any other operation, couldn't it?

Comment (oli-obk) [inline src/operator.rs]: Indeed, I'll rephrase it.

**Comment by oli-obk** [comment]
> Btw, I assume this also rules out pointer arithmetic? (Like the stuff I did to make HashMap work.)

Yea, that definitely can't work right now.

**Pull request merged**